### DISCUSSION TITLE
RIFE scale=0.5 not improving performance

### DISCUSSION THREAD

**Discussion by Chrysanthemum536** [AmusementClub/vs-mlrt#103] [Q&A] [answered]
is it just me or does setting scale=0.5 not help or even slightly decreases performance compared to scale=1? If I downscale the video myself before frame generation the performance does increase considerably, although I suppose that's not the same thing. I am running rife 4.4 using tensort.

**Comment by WolframRhodium** [comment]
Do you have benchmark data?

**Reply by Chrysanthemum536** [reply]
How should I benchmark? I am just using as a filter for mpv right now and seeing what speed I can go to before dropping frames.

**Reply by WolframRhodium** [reply]
Fine, then I'll do the benchmark myself. What's the video resolution and hardware configuration?

**Reply by Chrysanthemum536** [reply]
I am using nvidia 3090 and tried both 1440p and 2160p 24fps upscale to 48fps

**Reply by WolframRhodium** [reply, marked as answer]
A 3090 may not capable to do that. My benchmark with vsmlrt v15.2 on a 3090 using a 2880x1472 input:
| config | fps |
|:------:|:----|
| `_implementation=1, scale=0.5` | 44.18 |
| `_implementation=1, scale=1` | 41.18 |
| `_implementation=2, scale=1` | 58.84 |

(`_implementation=2` only supports `scale=1`, and will use `_implementation=1` as a fallback for other scale values.)

According to the data, scaling barely improves performance, but the v2 implementation is much faster here. This benchmark is done in native VapourSynth on a server xeon processor (Xeon Gold 6226R). The mpv playback also come with overheads.

**Reply by Chrysanthemum536** [reply]
ok, thanks for confirming.

i did not have _implementation set, so probably it was using implementation 2 without scaling and implementation 1 with.

**Comment by CANNONGH** [comment]
Hello, thank you very much for this great project.

I'm currently trying to use RIFE with an RTX 4070 Ti to interpolate 4K 30fps video to 60fps. I've been testing this using mpv-lazy, but due to the high resolution of the input, the process is too heavy to complete frame interpolation in real time—GPU usage becomes maxed out, and playback stutters or fails altogether.

While looking for optimization methods, I came across the following command in the [Practical-RIFE](https://github.com/hzwer/Practical-RIFE) repository:

`python3 inference_video.py --multi=2 --video=video.mp4 --scale=0.5`
I thought using --scale=0.5 would be a good way to reduce the processing load. However, in the file vsmlrt.py around lines 1036–1037, I found the following check:

```
if (model_major, model_minor) >= (4, 7) and scale != 1.0:
    raise ValueError("not supported")
```
This seems to indicate that --scale is not supported in RIFE version 4.7 and later.
I couldn’t find any mention of this limitation in the README or documentation—perhaps I missed it, but if not, I’d like to understand the reason behind this restriction in the code.

Additionally, if there are any alternative approaches or recommended settings for interpolating 4K videos efficiently with an RTX 4070 Ti, I would really appreciate any suggestions.

Thank you in advance!

**Reply by WolframRhodium** [reply]
Hi, thank you for your interest.

The reason that it is not supported is because of the engineering complexity of implementing that feature in ONNX, the format for RIFE models consumed by vs-mlrt. There is indeed no theoretical limitation on supporting this.

In particular, for RIFE v4.6 and earlier models, I used an older conversion script from PyTorch to ONNX which involves a lot of manual steps and is specific to a particular PyTorch version with custom modifications. For newer RIFE models, I use another conversion script that eliminates as much manual modifications as possible while being performant, and supporting this feature is usually version-specific.

Sorry for the inconvenience.

**Reply by CANNONGH** [reply]
Thank you for the detailed explanation — that makes a lot of sense and clears things up.

I have a follow-up question regarding model performance.
For example, I noticed that RIFE 4.25 lite has a tilesize_requirement of 128, which is double that of the standard models (usually 64).
Is this change in any way related to the scale=0.5 approach that was used in RIFE versions 4.6 and earlier to improve performance?
Does this imply that the lite model is approximately twice as lightweight in terms of GPU load or processing requirements? Or is the relationship more nuanced?

Thanks again for your time and help!

**Reply by WolframRhodium** [reply]
> Is this change in any way related to the scale=0.5 approach that was used in RIFE versions 4.6 and earlier to improve performance?

I cannot provide an authoritative answer to this question. This question should be asked in the upstream repository (https://github.com/hzwer/Practical-RIFE).

> Does this imply that the lite model is approximately twice as lightweight in terms of GPU load or processing requirements? Or is the relationship more nuanced?

I guess the relationship is more nuanced in practice. I do not have first-hand benchmark data for this.

**Reply by CANNONGH** [reply]
Thank you very much for your reply and clarification — I really appreciate it.
I see, so the relationship is a bit more nuanced than a simple comparison.
I'll check with the upstream repository as you suggested.

Thanks again for your time and help!
